### DISCUSSION TITLE
How to build a .ecore model from scratch?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#524] [Q&A] [answered]
Hi,
I tried to isolate the glsp-workflow-example into a separate repo to get a better understanding about all the concepts behind. (https://github.com/rsoika/glsp-workflow-example).

Building the glsp-workflow-server and starting the diagram.html from the workflow-standalone code base in a browser works all fine so far.

But now I tried t understand how to bring my own model. My expectation is, that I define my .ecore file in the model folder and than I choose from the context menu of the EMF file:

_New->Other->EMF Generator Model_

But then when I try to load the model, the Eclipse EMF Generator shows a validation error:

![image](https://user-images.githubusercontent.com/1339124/152681117-18ad5850-5419-414b-82d3-7844bf664bd1.png)

The reason for that validation is (as far as I understand for now) the missing `eSuperTypes`

![image](https://user-images.githubusercontent.com/1339124/152680904-076c4584-c7e0-43f0-b6b8-5b61dc72131b.png)

These types are references to the glsp-server code. How can I define those reference in case my code repo does not include the org.eclipse.glsp.graph model ?

**Comment by rsoika** [comment, marked as answer]
I figured out that I can copy the glsp-graph.ecore file into my project and I tweaked the .ecore file by changing the eSuperType path. This helped to generate my classes.

See my Documentation here: 

https://github.com/rsoika/glsp-workflow-example/tree/master/glsp-workflow-example-server#bring-your-own-model-extending-glsp-using-an-external-emf-model

Would it make sense to provide some kind of documentation also in the glsp project? Or is this a kind of competition in reverse engeneering ;-)

**Reply by planger** [reply]
Thanks for the feedback and providing a documented solution!
There will be a dedicated documentation page very soon. Here is the [preview](https://deploy-preview-45--glsp.netlify.app/documentation/) from the [PR](https://github.com/eclipsesource/graphical-lsp-website/pull/45).

It contains a short section about [extending the GModel](https://deploy-preview-45--glsp.netlify.app/documentation/modelextension/). But it is not as elaborated as your documentation. So please feel free to open a PR to add your documentation, once the PR linked above is merged.

It is worth noting though that extending the GModel is really something that is rarely required. Most of the times Args and CSS classes are sufficient in the projects I'm involved in.

**Reply by rsoika** [reply]
Thanks for your feedback. When you say "_extending the GModel is really something that is rarely required_", did you mean working with an EMF model and a .ecore file is rarely required? Or did I misunderstand you here? You know I plan to build a BPMN modeler and starting with an EMF model seems to me to be the right approach?

**Reply by planger** [reply]
Yes sure, you'd start from an Ecore and create (or use the existing) BPMN Ecore + generated Java classes to load your BPMN model. But then you take your BPMN model instance at runtime when the user opens the diagram and _transform_ it into a GModel, which is then sent to the client and rendered there.
The GModel is not your actual source model (your BPMN is). The GModel is just the description of the diagram to be rendered. So typically there is no need to extend the metamodel of the GModel.
Maybe this video helps: https://www.youtube.com/watch?v=tpCY4oJgor8

**Reply by rsoika** [reply]
Wow - yes this documentation preview is the piece I missed! I think will is answering a lot of my questions. I tried to get answers this weekend by reverse engineering the workflow-example which is very very arduous ;-)

I am already beginning to make some progress. Now work on the GLSP Server part of my BPMN model. The client part still worries me the most, because I am not very familiar with JavaScript. But with the new documentation it will probably be easier.

**Reply by planger** [reply]
We've merged the documentation today and it is available online: https://www.eclipse.org/glsp/documentation
